src/coreclr/jit/fgbasic.cpp

```diff
@@ -4618,6 +4618,50 @@ BasicBlock* Compiler::fgSplitBlockAfterStatement(BasicBlock* curr, Statement* st
     return newBlock;
 }
 
+//------------------------------------------------------------------------------
+// fgSplitBlockBeforeTree : Split the given block right before the given tree
+//
+// Arguments:
+//    block        - The block containing the statement.
+//    stmt         - The statement containing the tree.
+//    splitPoint   - A tree inside the statement.
+//    firstNewStmt - [out] The first new statement that was introduced.
+//                   [firstNewStmt..stmt) are the statements added by this function.
+//    splitNodeUse - The use of the tree to split at.
+//
+// Returns:
+//    The last block after split
+//
+// Notes:
+//    See comments in gtSplitTree
+//
+BasicBlock* Compiler::fgSplitBlockBeforeTree(
+    BasicBlock* block, Statement* stmt, GenTree* splitPoint, Statement** firstNewStmt, GenTree*** splitNodeUse)
+{
+    gtSplitTree(block, stmt, splitPoint, firstNewStmt, splitNodeUse);
+
+    BasicBlockFlags originalFlags = block->bbFlags;
+    BasicBlock*     prevBb        = block;
+
+    if (stmt == block->firstStmt())
```

Comment:
I'll add a comment in a follow up if you don't mind to avoid re-running CI for it

Reply:
Is this really about `gtSplitTree`? Isn't it here because we don't have a `fgSplitBlockBefore`?

`gtSplitTree` can make changes even without introducing new statements -- the return value needs to be used for that kind of check.

Reply:
Yes, it's here just because I need basically `fgSplitBlockBefore` and only have `fgSplitBlockAfter` API

Reply:
so when it's not the first statement in the current block I do `fgSplitBlockAfter (stmt->PrevStmt())`

Reply:
I suppose I should look at what `gtSplitTree` does, but I'm still confused exactly what this method is supposed to be doing. Maybe an example would help?

Reply:
> I suppose I should look at what `gtSplitTree` does, but I'm still confused exactly what this method is supposed to be doing. Maybe an example would help?

The general idea that we want to split a block (e.g. `BB0`) into two (say, `BBa` and `BBb`) at a specific node (e.g. `callX`) and make sure that all side-effects are moved to `BBa` and the actual `callX` is now in `BBb`.

We have a phase where we insert GC safe points after specific `call` nodes - that one didn't need to take care about any execution ordering since we just wanted to make sure GC is polled (it doesn't even matter whether to emit the poll before or after the calls). The runtime lookup case is a lot more complicated since we needed to insert verbose tree in front of the call and re-use its arguments, respect all kinds of complex COMMAs,